docs/features/communication/abi_compatible_data_types/index.rst

```diff
@@ -0,0 +1,491 @@
+
+Assumptions
+^^^^^^^^^^^
+
+* Shared memory regions are mapped at correctly aligned virtual addresses in both processes.
+* No serialization or runtime copying occurs when interpreting a type from memory.
+* Processes use the same endianness.
+* No synchronization or atomicity guarantees are defined at the data type level; these are provided by the IPC framework.
+* All memory is allocated statically or pre-reserved. Dynamic memory allocation is disallowed.
+
+Type Conformance
+^^^^^^^^^^^^^^^^
+
+Types used in shared memory must meet the following criteria:
+
+1. **Fixed size and alignment**: Every type must have a known, constant size and alignment at compile time.
+2. **Consistent layout across languages**: The layout of a type must be identical in Rust and C++, and on all platforms.
+3. **No absolute pointers or references**: Types must not contain pointers to heap memory, function pointers, or references.
+4. **No language-specific metadata**: No vtables, slice headers, or implementation-specific type markers are allowed.
+
+Each type definition must clearly indicate whether it conforms to these rules natively or requires a custom definition to do so.
+
+Type Categories
+^^^^^^^^^^^^^^^
+
+Primitive Types
+"""""""""""""""
+
+These types are ABI-compatible when declared using fixed-size standard types:
+
+
+.. list-table:: Native Type Mapping
+   :header-rows: 1
+
+   * - Concept
+     - Rust
+     - C++17
+   * - Boolean
+     - ``bool``
+     - ``bool`` (1 byte, with ``0x00`` and ``0x01`` as the only valid bit patterns)
+   * - Integers (N = 8, 16, 32, 64)
+     - ``uN``, ``iN``
+     - ``std::uintN_t``, ``std::intN_t``
+   * - Floating point
+     - ``f32``, ``f64``
+     - ``float``, ``double``
+
+All types must avoid trap representations and undefined padding.
+
+Structs and Tuples
+""""""""""""""""""
+
+Structs and tuples are supported using standard layout rules:
+
+* **Rust**: Requires ``#[repr(C)]``
+  (guarantees C-compatible memory layout;
+  `full specification <https://doc.rust-lang.org/reference/type-layout.html#the-c-representation>`__)
+* **C++**: Requires ``standard_layout``
+  (no virtual functions, no virtual inheritance, and only one class in the hierarchy has non-static data members;
+  `full specification <https://en.cppreference.com/w/cpp/language/classes.html#Standard-layout_class>`__)
+
+Field ordering must be preserved and padding must be identical across compilers. Any alignment greater than the default must be explicitly declared.
+
+Enums
+"""""
+
+Only fieldless enums with a defined underlying integer type are supported. These must use:
+
+* ``#[repr(u8)]``, ``#[repr(u16)]``, etc. in Rust
+* ``enum class MyEnum : std::uint8_t`` in C++
+
+Enums with payloads (discriminated unions) are not supported.
```

Comment:
We could mention that a variant can be used to work around this limitation?

Not a must from my side, just an idea.